### PR TITLE
Include data source type in the logical plan

### LINKED ISSUE #1637: RUN SCRIPT and non-interactive mode incorrectly parses S-T join as S-S and aborts
Version: 5.0.0-rc3 (Docker)

Script seems (c.f #1638 - it doesn't actually say) to barf out at the point of join a join between PAGEVIEWS_TRANSFORMED (stream) and USERS_5PART (table). This SQL is valid and runs fine when entered manually in the CLI.

Happens with both `RUN SCRIPT` in CLI, and `--queries-file` in ksql-server

Run from ksql-server: 

	docker run \
	    -v $PWD/examples.sql:/data/examples.sql \
	    --network tutorials_default \
	    -e KSQL_BOOTSTRAP_SERVERS=kafka:39092 \
	    -e KSQL_KSQL_SERVICE_ID=rmoff_test_01 \
	    -e KSQL_KSQL_QUERIES_FILE=/data/examples.sql \
	    -e KSQL_KSQL_SCHEMA_REGISTRY_URL=http://schema-registry:8081 \
	    confluentinc/cp-ksql-server:5.0.0-rc3

Error from ksql-server: 

	[2018-07-26 16:28:23,798] ERROR Failed to start KSQL Server with query file: /data/examples.sql (io.confluent.ksql.rest.server.StandaloneExecutor:70)
	io.confluent.ksql.util.KsqlException: Stream-Stream joins must have a WITHIN clause specified. None was provided. To learn about how to specify a WITHIN clause with a stream-stream join, please visit: https://docs.confluent.io/current/ksql/docs/syntax-reference.html#create-stream-as-select
		at io.confluent.ksql.planner.plan.JoinNode$StreamToStreamJoiner.join(JoinNode.java:356)
		at io.confluent.ksql.planner.plan.JoinNode.buildStream(JoinNode.java:187)
		at io.confluent.ksql.planner.plan.ProjectNode.buildStream(ProjectNode.java:124)
		at io.confluent.ksql.planner.plan.KsqlStructuredDataOutputNode.buildStream(KsqlStructuredDataOutputNode.java:100)
		at io.confluent.ksql.physical.PhysicalPlanBuilder.buildPhysicalPlan(PhysicalPlanBuilder.java:106)
		at io.confluent.ksql.QueryEngine.buildQueryPhysicalPlan(QueryEngine.java:199)
		at io.confluent.ksql.QueryEngine.buildPhysicalPlans(QueryEngine.java:165)
		at io.confluent.ksql.KsqlEngine.planQueries(KsqlEngine.java:228)
		at io.confluent.ksql.KsqlEngine.createQueries(KsqlEngine.java:618)
		at io.confluent.ksql.rest.server.StandaloneExecutor.executeStatements(StandaloneExecutor.java:118)
		at io.confluent.ksql.rest.server.StandaloneExecutor.start(StandaloneExecutor.java:67)
		at io.confluent.ksql.rest.server.KsqlServerMain.tryStartApp(KsqlServerMain.java:60)
		at io.confluent.ksql.rest.server.KsqlServerMain.main(KsqlServerMain.java:46)
	[2018-07-26 16:28:23,814] INFO Server shutting down (io.confluent.ksql.rest.server.KsqlServerMain:64)

Error from CLI: 

	ksql> RUN SCRIPT '/data/examples.sql';
	io.confluent.ksql.util.KsqlException: Stream-Stream joins must have a WITHIN clause specified. None was provided. To learn about how to specify a WITHIN clause with a stream-stream join, please visit: https://docs.confluent.io/current/ksql/docs/syntax-reference.html#create-stream-as-select
	ksql> show streams;

	 Stream Name                      | Kafka Topic                      | Format
	---------------------------------------------------------------------------------
	 PAGEVIEWS_TRANSFORMED            | PAGEVIEWS_TRANSFORMED            | JSON
	 PAGEVIEWS_TRANSFORMED_PRIORITY_1 | PAGEVIEWS_TRANSFORMED_PRIORITY_1 | JSON
	 PAGEVIEWS_TRANSFORMED_PRIORITY_2 | PAGEVIEWS_TRANSFORMED_PRIORITY_2 | JSON
	 PAGEVIEWS                        | pageviews                        | DELIMITED
	---------------------------------------------------------------------------------
	ksql> show tables;

	 Table Name  | Kafka Topic | Format | Windowed
	-----------------------------------------------
	 USERS_5PART | USERS_5PART | JSON   | false
	 USERS       | users       | JSON   | false
	-----------------------------------------------

Source SQL: 

```
    CREATE STREAM pageviews 
      (viewtime BIGINT, 
       userid VARCHAR, 
       pageid VARCHAR) 
      WITH (KAFKA_TOPIC='pageviews', 
            VALUE_FORMAT='DELIMITED');
    CREATE TABLE users 
      (registertime BIGINT, 
       gender VARCHAR, 
       regionid VARCHAR, 
       userid VARCHAR, 
       interests array<VARCHAR>, 
       contactinfo map<VARCHAR, VARCHAR>) 
      WITH (KAFKA_TOPIC='users', 
            VALUE_FORMAT='JSON', 
            KEY = 'userid');
    CREATE STREAM pageviews_transformed 
      WITH (TIMESTAMP='viewtime', 
            PARTITIONS=5, 
            VALUE_FORMAT='JSON') AS 
      SELECT viewtime, 
             userid, 
             pageid, 
             TIMESTAMPTOSTRING(viewtime, 'yyyy-MM-dd HHmmss.SSS') AS timestring 
      FROM pageviews 
      PARTITION BY userid;
    CREATE STREAM pageviews_transformed_priority_1 
      WITH (TIMESTAMP='viewtime', 
            PARTITIONS=5, 
            VALUE_FORMAT='JSON') AS 
      SELECT viewtime, 
             userid, 
             pageid, 
             TIMESTAMPTOSTRING(viewtime, 'yyyy-MM-dd HHmmss.SSS') AS timestring 
      FROM pageviews 
      WHERE userid='User_1' OR userid='User_2' 
      PARTITION BY userid;
    CREATE STREAM pageviews_transformed_priority_2 
          WITH (TIMESTAMP='viewtime', 
                PARTITIONS=5, 
                VALUE_FORMAT='JSON') AS 
      SELECT viewtime, 
             userid, 
             pageid, 
             TIMESTAMPTOSTRING(viewtime, 'yyyy-MM-dd HHmmss.SSS') AS timestring 
      FROM pageviews 
      WHERE userid<>'User_1' AND userid<>'User_2' 
      PARTITION BY userid;
    CREATE TABLE users_5part 
        WITH (PARTITIONS=5) AS 
        SELECT * FROM USERS;
    CREATE STREAM pageviews_enriched AS 
      SELECT pv.viewtime, 
             pv.userid AS userid, 
             pv.pageid, 
             pv.timestring, 
             u.gender, 
             u.regionid, 
             u.interests, 
             u.contactinfo 
      FROM pageviews_transformed pv 
      LEFT JOIN users_5part u ON pv.userid = u.userid;
    CREATE TABLE pageviews_per_region AS 
      SELECT regionid, 
             count(*) 
      FROM pageviews_enriched 
      GROUP BY regionid;
    CREATE TABLE pageviews_per_region_per_minute AS 
      SELECT regionid, 
             count(*) 
      FROM pageviews_enriched 
      WINDOW TUMBLING (SIZE 1 MINUTE) 
      GROUP BY regionid;
   CREATE TABLE pageviews_per_region_per_30secs AS 
      SELECT regionid, 
             count(*) 
      FROM pageviews_enriched 
      WINDOW TUMBLING (SIZE 30 SECONDS) 
      WHERE UCASE(gender)='FEMALE' AND LCASE(regionid)='region_6' 
      GROUP BY regionid;
   CREATE TABLE pageviews_per_region_per_30secs10secs AS 
      SELECT regionid, 
             count(*) 
      FROM pageviews_enriched 
      WINDOW HOPPING (SIZE 30 SECONDS, ADVANCE BY 10 SECONDS) 
      WHERE UCASE(gender)='FEMALE' AND LCASE (regionid) LIKE '%_6' 
      GROUP BY regionid;
   CREATE TABLE pageviews_per_region_per_session AS 
      SELECT regionid, 
             count(*) 
      FROM pageviews_enriched 
      WINDOW SESSION (60 SECONDS) 
      GROUP BY regionid;
    CREATE STREAM pageviews_interest_contact AS 
   SELECT interests[0] AS first_interest, 
             contactinfo['zipcode'] AS zipcode, 
             contactinfo['city'] AS city, 
             viewtime, 
             userid, 
             pageid, 
             timestring, 
             gender, 
             regionid 
      FROM pageviews_enriched;
```

### PR DESCRIPTION
**Pull request by hjafarpour**
### Description 
This will make sure that the logical plan nodes have correct data source type.
This fixes #1637 

### Testing done 
Unit tests were added.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 1]
```diff
@@ -80,7 +81,9 @@ public JoinNode(@JsonProperty("id") final PlanNodeId id,
                   @JsonProperty("rightType") final DataSource.DataSourceType rightType) {
 
     // TODO: Type should be derived.
-    super(id);
+    super(id, (leftType == DataSourceType.KTABLE && rightType == DataSourceType.KTABLE)
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java]: Can't this just be simplified to leftType ?

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 2]
```diff
@@ -188,4 +189,72 @@ public void testComplexAggregateLogicalPlan() {
     assertThat(logicalPlan.getSources().get(0).getSchema().fields().size(), equalTo(2));
 
   }
+
+  @Test
+  public void shouldCreateTableOutputForAggregateQuery() {
+    String simpleQuery = "SELECT col0, sum(floor(col3)*100)/count(col3) FROM test1 window "
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: final here and below

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java (hunk 1 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java]: Yes, but this is more stricter check!

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 3]
```diff
@@ -188,4 +189,72 @@ public void testComplexAggregateLogicalPlan() {
...
+        + "WHERE col0 > 100 GROUP BY col0;";
+
+    PlanNode logicalPlan = buildLogicalPlan(simpleQuery);
+    assertThat(logicalPlan.getNodeOutputType(), equalTo(DataSourceType.KTABLE));
+    System.out.println();
+  }
+
+  @Test
+  public void shouldCreateStreamOutputForStreamTableJoin() {
+    String
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: simpleQuery and logicalPlan should be final

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 4]
```diff
@@ -188,4 +189,72 @@ public void testComplexAggregateLogicalPlan() {
...
+        simpleQuery =
+        "SELECT t1.col1, t2.col1, col5, t2.col4, t2.col2 FROM test1 t1 LEFT JOIN test2 t2 ON "
+            + "t1.col1 = t2.col1 WHERE t1.col1 > 10 AND t2.col4 = 10.8;";
+    PlanNode logicalPlan = buildLogicalPlan(simpleQuery);
+    assertThat(logicalPlan.getNodeOutputType(), equalTo(DataSourceType.KSTREAM));
+  }
+
+  @Test
+  public void shouldCreateStreamOutputForStreamFilter() {
+    String
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: simpleQuery and logicalPlan should be final

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 5]
```diff
@@ -188,4 +189,72 @@ public void testComplexAggregateLogicalPlan() {
...
+        simpleQuery = "SELECT * FROM test1 WHERE col0 > 100;";
+    PlanNode logicalPlan = buildLogicalPlan(simpleQuery);
+    assertThat(logicalPlan.getNodeOutputType(), equalTo(DataSourceType.KSTREAM));
+  }
+
+  @Test
+  public void shouldCreateTableOutputForTableFilter() {
+    String
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: simpleQuery and logicalPlan should be final

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 6]
```diff
@@ -188,4 +189,72 @@ public void testComplexAggregateLogicalPlan() {
...
+        simpleQuery = "SELECT * FROM test2 WHERE col4 = 10.8;";
+    PlanNode logicalPlan = buildLogicalPlan(simpleQuery);
+    assertThat(logicalPlan.getNodeOutputType(), equalTo(DataSourceType.KTABLE));
+  }
+
+  @Test
+  public void shouldCreateStreamOutputForStreamProjection() {
+    String
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: simpleQuery and logicalPlan should be final

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 7]
```diff
@@ -188,4 +189,72 @@ public void testComplexAggregateLogicalPlan() {
...
+        simpleQuery = "SELECT col0 FROM test1;";
+    PlanNode logicalPlan = buildLogicalPlan(simpleQuery);
+    assertThat(logicalPlan.getNodeOutputType(), equalTo(DataSourceType.KSTREAM));
+  }
+
+  @Test
+  public void shouldCreateTableOutputForTableProjection() {
+    String
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: simpleQuery and logicalPlan should be final

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 8]
```diff
@@ -188,4 +189,72 @@ public void testComplexAggregateLogicalPlan() {
...
+        simpleQuery = "SELECT col4 FROM test2;";
+    PlanNode logicalPlan = buildLogicalPlan(simpleQuery);
+    assertThat(logicalPlan.getNodeOutputType(), equalTo(DataSourceType.KTABLE));
+  }
+
+  @Test
+  public void shouldCreateStreamOutputForStreamStreamJoin() {
+    String simpleQuery = "SELECT * FROM ORDERS INNER JOIN TEST1 ON ORDERS.ORDERID=TEST1.COL0;";
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: simpleQuery and logicalPlan should be final

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 9]
```diff
@@ -188,4 +189,72 @@ public void testComplexAggregateLogicalPlan() {
...
+    PlanNode logicalPlan = buildLogicalPlan(simpleQuery);
+    assertThat(logicalPlan.getNodeOutputType(), equalTo(DataSourceType.KSTREAM));
+  }
+
+  @Test
+  public void shouldCreateTableOutputForTableTableJoin() {
+    String simpleQuery = "SELECT * FROM TEST2 INNER JOIN TEST3 ON TEST2.COL0=TEST3.COL0;";
```
Comment (rodesai) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: simpleQuery and logicalPlan should be final

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java [hunk 10]
```diff
@@ -186,4 +186,72 @@ public void testComplexAggregateLogicalPlan() {
...
+
+  @Test
+  public void shouldCreateTableOutputForAggregateQuery() {
+    final String simpleQuery = "SELECT col0, sum(floor(col3)*100)/count(col3) FROM test1 window "
+        + "HOPPING ( size 2 second, advance by 1 second) "
+        + "WHERE col0 > 100 GROUP BY col0;";
+
+    final PlanNode logicalPlan = buildLogicalPlan(simpleQuery);
+    assertThat(logicalPlan.getNodeOutputType(), equalTo(DataSourceType.KTABLE));
+    System.out.println();
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: this should be removed.

File: ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java [hunk 11]
```diff
@@ -115,17 +117,28 @@ private PlanNode buildQueryLogicalPlan(
...
           structuredDataSource =
-          new KsqlStream(
-              sqlExpression,
-              ksqlStructuredDataOutputNode.getId().toString(),
-              ksqlStructuredDataOutputNode.getSchema(),
-              ksqlStructuredDataOutputNode.getKeyField(),
-              ksqlStructuredDataOutputNode.getTimestampExtractionPolicy(),
-              ksqlStructuredDataOutputNode.getKsqlTopic()
-          );
+          (ksqlStructuredDataOutputNode.getNodeOutputType() == DataSourceType.KTABLE)
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java]: For my own understanding, how is this working exactly? The `structuredDataSource` object seems to only be used if `analysis.isDoCreateInto` is `true`, which only seems to be the case for `INSERT INTO` statements. This is the only place where `getNodeOutputType` is called. So how does this change the behavior for non `INSERT INTO` statements?

Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java]: The opposite, indeed `isDoCreateInto` is true only for CSAS/CTAS and it's false for `INSERT INTO`.  For `INSERT INTO` we already have the output stream and no need to set it's type!

File: ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java (hunk 10 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/planner/LogicalPlannerTest.java]: Good catch! removed it.

File: ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java (hunk 11 above)
Comment (apurvam) [reply ksql-engine/src/main/java/io/confluent/ksql/QueryEngine.java]: ah. thanks for the clarification.

**Pull request merged**